pkg/network/driver/nmstate/types.go

```diff
@@ -0,0 +1,233 @@
+	"kubevirt.io/kubevirt/pkg/network/driver/procsys"
+	"kubevirt.io/kubevirt/pkg/network/driver/virtchroot"
+)
+
+type Spec struct {
+	Interfaces []Interface `json:"interfaces,omitempty"`
+	LinuxStack LinuxStack  `json:"linux-stack,omitempty"`
+}
+
+type Status struct {
+	Interfaces []Interface `json:"interfaces"`
+	Routes     Routes      `json:"routes"`
+	LinuxStack LinuxStack  `json:"linux-stack,omitempty"`
+}
+
+type Interface struct {
+	Name        string  `json:"name"`
+	Index       int     `json:"index,omitempty"`
+	TypeName    string  `json:"type,omitempty"`
+	State       string  `json:"state,omitempty"`
+	MacAddress  string  `json:"mac-address,omitempty"`
+	CopyMacFrom string  `json:"copy-mac-from,omitempty"`
+	MTU         int     `json:"mtu,omitempty"`
+	Controller  string  `json:"controller,omitempty"`
+	Ethtool     Ethtool `json:"ethtool,omitempty"`
+
+	Tap *TapDevice `json:"tap,omitempty"`
+
+	IPv4 IP `json:"IPv4,omitempty"`
+	IPv6 IP `json:"IPv6,omitempty"`
+
+	LinuxStack LinuxIfaceStack `json:"linux-stack,omitempty"`
+
+	Metadata *IfaceMetadata
+}
+
+type IP struct {
+	Enabled *bool       `json:"enabled,omitempty"`
+	Address []IPAddress `json:"address,omitempty"`
+}
+
+type IPAddress struct {
+	IP        string `json:"IP,omitempty"`
+	PrefixLen int    `json:"prefix-length,omitempty"`
+}
+
+type Routes struct {
+	Running []Route `json:"running,omitempty"`
+}
+
+type Route struct {
+	Destination      string `json:"destination"`
+	NextHopInterface string `json:"next-hop-interface,omitempty"`
+	NextHopAddress   string `json:"next-hop-address,omitempty"`
+	TableID          int    `json:"table-id,omitempty"`
+}
+
+type Ethtool struct {
+	Feature Feature `json:"feature,omitempty"`
+}
+
+type Feature struct {
+	TxChecksum *bool `json:"tx-checksum,omitempty"`
+}
+
+type TapDevice struct {
+	Queues int `json:"queues,omitempty"`
+	UID    int `json:"UID,omitempty"`
+	GID    int `json:"GID,omitempty"`
+}
+
+type LinuxIfaceStack struct {
+	IP4RouteLocalNet *bool `json:"ip4-route-local-net,omitempty"`
+	PortLearning     *bool `json:"port-learning,omitempty"`
+}
+
+type LinuxStack struct {
+	IPv4 LinuxStackIP4 `json:"ipv4,omitempty"`
+	IPv6 LinuxStackIP6 `json:"ipv6,omitempty"`
+}
+
+type LinuxStackIP4 struct {
+	ArpIgnore             *procsys.ArpReplyMode `json:"arp-ignore,omitempty"`
+	Forwarding            *bool                 `json:"forwarding,omitempty"`
+	PingGroupRange        []int                 `json:"ping-group-range,omitempty"`
+	UnprivilegedPortStart *int                  `json:"unprivileged-port-start,omitempty"`
+}
+
+type LinuxStackIP6 struct {
+	Forwarding *bool `json:"forwarding,omitempty"`
+}
+
+// IfaceMetadata includes extra data which is piggyback on the nmstate object.
+// Users of the nmstate object can use it to store data and use it in some scenarios (e.g. creating the tap device).
+type IfaceMetadata struct {
+	// Pid refers to the process ID of the virt-launcher.
+	Pid int
```

Comment:
Both the pid and the network name are very coupled to kubevirt.
Since the nmstate package is independent and maybe even be moved outside of kubevirt I would add a key value map as a metadata instead of concrete fields.